### DISCUSSION TITLE
Some questions about understanding and customizing Auto3dSeg

### DISCUSSION THREAD

**Discussion by vokcow** [Project-MONAI/tutorials#1001] [Q&A]
Hello, 

First of all, many thanks for Auto3dSeg, it is an amazing new feature of MONAI. I have some questions:

1.   In the tutorials it is shown how we can overwrite training parameters as

  
  ```
    runner = AutoRunner(
        work_dir=work_dir,
        input={
            "modality": "CT",
            "datalist": datalist_file,
            "dataroot": dataroot_dir,
        }
    )
      runner.set_training_params(...)
  ```
  
  Say I would like to use the same (custom) loss function for all the algos, how do I specify that? Is it trhoguh set_training_params() method?

2. I also note that the runner takes the datalist_file, which from the tutorials I see it is a dictionary of the following form:

```
{"testing":[{"image":"test_image_1.nii.gz", ...}], 
"training":[{"image":"train_image_1.nii.gz", "label":"train_label_1", "fold":0},...]}
```
Therefore, I would assume we can specify the cross validation folds from this datalist file, as for each training sample a fold is specified. However, it seems that the number of splits is determined by AutoRunner() and set to a default of 5, and modificable via the method 
```
    runner.set_num_fold(num_fold=...)
```
What is the purpose of the "fold" fields in the datalist file then? 

3. Finally, I notice that by default Auto3DSeg will try 5 different models, namely: Dints, Segresnet, Segresnet2d, Swinunet, and UNet, but I assume we should be able to choose which models to try? How can we set up which models to try and can we use our own custom models (custom code)? 

Many thanks in advance, 

Victor

**Comment by mingxin-zheng** [comment]
Hi @VokCow , 

1. About the loss function, I think it should be done via the `set_train_params`. Let me check the detailed usage and update you later.
2. It is a good point and I think it lacks some documentation in the tutorial/docstring. The datalist file has made 5 splits and the training will execute the first `num_fold` of the splits. For example, you can set it to 1-5 but can't set it to 6.
3. Custom model and code can be added but it is tedious. Tutorial is still in preparation, and here is an "unproper" example how you can add your models : https://github.com/Project-MONAI/research-contributions/blob/main/auto3dseg/tests/test_algo_templates.py#L127-L159
